doc/monitoring_containers/monitoring_containers.md

```diff
@@ -0,0 +1,342 @@
+container stops, the systemd service which manages the container will also stop, but it is
+configured to automatically restart the service, thus it will restart the container.
+
+# 2 Functionality
+## 2.1 Target Deployment Use Cases
+These two features are used to perform the following functions:
+1. Monit will write an alert message into syslog if one if critical process has not been
+    alive for 5 minutes.
+2. Monit will write an alert message into syslog if the usage of memory is larger than the
+    pre-defined threshold for a docker container.
+3. A docker container will auto-restart if one of its critical processes crashed or exited
+    unexpectedly.
+
+## 2.2 Functional Description
+
+
+### 2.2.1 Monitoring Critical Processes
+Monit natively implements a mechanism to monitor whether a process is running or not. In detail,
+Monit will periodically read the target processes from configuration file and try to match 
+those process with the processes tree in Linux kernel.
+
+Below is an example of Monit configuration file to monitor the critical processes in lldp
+container.
+
+*/etc/monit/conf.d/monit_lldp*
+```bash
+###############################################################################
+# Monit configuration file for lldp container
+# Process list:
+#   lldpd
+#   lldp_syncd
+#   lldpmgrd
+###############################################################################
+check process lldp_monitor matching "lldpd: "
+    if does not exit for 5 times within 5 cycles then alert
+check process lldp_syncd matching "python2 -m lldp_syncd"
+    if does not exit for 5 times within 5 cycles then alert
+check process lldpmgrd matching "python /usr/bin/lldpmgrd"
+    if does not exit for 5 times within 5 cycles then alert
+```
+However, Monit is unable to monitor multiple processes executing the same command but with
+different arguments. For example, in teamd container, there are multiple teamd processes 
+running the same command ```/usr/bin/teamd``` but using different port channel as argument.
+Since there exists 1:1 mapping between a port channel and a teamd process, we employ Monit to 
+monitor a script which retrieves all the port channels from Config_DB and then determine
+whether there exists a teamd process in Linux for each port channel. If succeed, that means
+all teamd processes are live. Otherwise, we will know at least teamd process exited unexpectedly
+and then Monit will write an alert message into syslog. Similarly we can also use this method
+to solve the same issue in dhcp_relay container.
+
+### 2.2.2 Monitoring Critical Resource Usage
+Similar to monitoring the critical processes, we can employ Monit to monitor the resource usage
+such as CPU, memory and disk for each process. Unfortunately Monit is unable to do the resource monitoring
+in the container level. Thus we propose a new design to achieve such monitoring based on Monit.
+Specifically Monit will monitor a script and check its exit status. This script
```

Comment:
Does this script be able to detect any hang/loop or deadlock situation for the processes or
threads inside the container?